### PR TITLE
Python3 support

### PR DESCRIPTION
**Pull request by RedCraig**
This PR adds support for python3, while maintaining support for python2.7.

Where possible I've used `try`/`except`:
```python
try:
    python2.7 code
except AttributeError:
    python3.5 code
```
In some cases I had to resort to checking the `sys.version_info` and determining the code path based on that.

Please pay extra close attention to the changes I made to `__GetAuthorizationTokenUsingMasterKey` as this is used in all request auth header generation.

Note: This was branched from the branch in PR #51, and therefore includes the changes made for PR #51.
# Unittests
### master test status:
`crud_tests.py` gives many `429` Request Rate too high errors, so it's difficult to fully test when branching/making changes. See #49.
- test/crud_tests.py
    Ran 72 tests, 33 errors:
  - `test_attachment_crud_self_link`:
      Fails intermittently/randomly. See Issue #52
  - `test_partitioned_collection_path_parser`:
    ```
      FileNotFoundError: [Errno 2] No such file or directory: '/Users/<name>/.net/Microsoft.Azure.Documents.Client.Test/Routing/resources/BaselineTest.PathParser.json
    ```
  - `test_id_validation`:
    ```
    {"code":"Conflict","message":"Message: {\\"Errors\\":[\\"Resource with specified id or name already exists\\"
    ```
  - `test_offer_replace`:
      ```pydocumentdb.errors.HTTPFailure: Status code: 400
      b'{"code":"BadRequest","message":"Offer type is not supported with Offer version \'V2\' and above.\\r\\nActivityId: f7758066-ddd7-4521-b72d-a220ce6ed37b"}'
      ```
  - `test_sql_query_crud`
      `HTTPFailure: Status code: 409`
      `{"code":"Conflict","message":"Message: {\"Errors\":[\"Resource with specified id or name already exists\"]}`
- python test/ttl_tests.py:
    pass

### python3 branch unittests using `python3.5` & `python2.7`:
- test/crud_tests.py - same as master.
- python test/ttl_tests.py - pass - same as master.

## HTTP Connection closure
Python3 warns:
```bash
/Projects/azure-documentdb-python/pydocumentdb/synchronized_request.py:179: 
ResourceWarning: unclosed <ssl.SSLSocket fd=4, 
family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, 
proto=6, laddr=(<addr ommitted>), raddr=(<addr ommitted>)>
```
the code now explicitly closes HTTP connections _except in the streaming case_. If the data is streamed then the response is passed back, then this needs some other mechanism for the user/API to close the connection when streaming is complete, as immediate closure won’t work.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@RedCraig__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, AZPRBOT;

File: pydocumentdb/base.py [hunk 1]
```diff
@@ -97,9 +103,13 @@ def GetHeaders(document_client,
         headers[http_constants.HttpHeaders.OfferThroughput] = options['offerThroughput']
 
     if 'partitionKey' in options:
-        # if partitionKey value is Undefined, serailize it as {} to be consistent with other SDKs
+        # if partitionKey value is Undefined, serialize it as {} to be consistent with other SDKs
         if options.get('partitionKey') is documents.Undefined:
-            headers[http_constants.HttpHeaders.PartitionKey] = [{}]
+            import sys
```
Comment (tomviner) [inline pydocumentdb/base.py]: `sys` is already imported.

File: test/crud_tests.py [hunk 2]
```diff
@@ -3,25 +3,34 @@
...
+import unittest
+import sys
 
-import pydocumentdb.documents as documents
+try:
+    # Python2.7
+    from __builtin__ import *
+except ImportError:
+    # Python3
+    import builtins
```
Comment (tomviner) [inline test/crud_tests.py]: What's the need for these imports?

If they are needed, should Python 3 be `from builtins import *` to match what's done for Python 2?

File: pydocumentdb/base.py (hunk 1 above)
Comment (rnagpal) [inline pydocumentdb/base.py]: Thanks for pointing out. We will clean this up as part of the Python3 conversion we are currently working on.

File: test/crud_tests.py (hunk 2 above)
Comment (rnagpal) [inline test/crud_tests.py]: We are working on Python 3 conversion and will update it as needed.

**Comment by rnagpal** [comment]
@RedCraig We have made lots of product changes and released 2.0.0 version which has the support for Python3, so I'm closing your Pull request. Let us know if you still see any issues running the tests with Python3.

**Pull request closed without merging**